File: python/roteiro/collections.py
"""OGC API Features collections and feature operations.

This module provides standalone functions that operate on a ``RoteiroClient``
instance.  They are also accessible as methods on ``RoteiroClient`` directly
(the client class delegates to these).
"""

from __future__ import annotations

from typing import TYPE_CHECKING, List, Optional

from .client import _encode_path_value, _with_query
from .models import Collection, Feature, FeatureCollection

if TYPE_CHECKING:
    from .client import RoteiroClient


def list_collections(client: RoteiroClient) -> List[Collection]:
    """List all OGC API collections.

    Args:
        client: An initialised RoteiroClient instance.

    Returns:
        A list of Collection descriptors.
    """
    resp = client._get("/collections")
    return [Collection.from_dict(c) for c in resp.get("collections", [])]


def get_collection(client: RoteiroClient, collection_id: str) -> Collection:
    """Get a single collection by ID.

    Args:
        client: An initialised RoteiroClient instance.
        collection_id: The collection identifier.

    Returns:
        A Collection object.
    """
    data = client._get(f"/collections/{_encode_path_value(collection_id)}")
    return Collection.from_dict(data)


def get_items(
    client: RoteiroClient,
    collection_id: str,
    *,
    bbox: Optional[str] = None,
    limit: Optional[int] = None,
    where: Optional[str] = None,
) -> FeatureCollection:
    """Query features from a collection with optional filters.

    Args:
        client: An initialised RoteiroClient instance.
        collection_id: The collection identifier.
        bbox: Bounding box filter as ``"minx,miny,maxx,maxy"``.
        limit: Maximum number of features to return.
        where: CQL2 filter expression.

    Returns:
        A FeatureCollection with matching features.
    """
    path = _with_query(
        f"/collections/{_encode_path_value(collection_id)}/items",
        [("bbox", bbox), ("limit", limit), ("filter", where)],
    )
    data = client._get(path)
    return FeatureCollection.from_dict(data)


def get_item(
    client: RoteiroClient,
    collection_id: str,
    feature_id: str,
) -> Feature:
    """Get a single feature from a collection.

    Args:
        client: An initialised RoteiroClient instance.
        collection_id: The collection identifier.
        feature_id: The feature identifier.

    Returns:
        A Feature object.
    """
    data = client._get(
        f"/collections/{_encode_path_value(collection_id)}/items/{_encode_path_value(feature_id)}"
    )
    return Feature.from_dict(data)


def create_item(
    client: RoteiroClient,
    collection_id: str,
    feature: dict,
) -> Feature:
    """Create a new feature in a collection.

    Args:
        client: An initialised RoteiroClient instance.
        collection_id: The collection identifier.
        feature: A GeoJSON Feature dictionary.

    Returns:
        The created Feature.
    """
    data = client._request(
        "POST",
        f"/collections/{_encode_path_value(collection_id)}/items",
        body=feature,
        extra_headers={"Content-Type": "application/geo+json"},
    )
    return Feature.from_dict(data)


def update_item(
    client: RoteiroClient,
    collection_id: str,
    feature_id: str,
    feature: dict,
) -> Feature:
    """Update an existing feature in a collection.

    Args:
        client: An initialised RoteiroClient instance.
        collection_id: The collection identifier.
        feature_id: The feature identifier.
        feature: A GeoJSON Feature dictionary with updated data.

    Returns:
        The updated Feature.
    """
    data = client._request(
        "PUT",
        f"/collections/{_encode_path_value(collection_id)}/items/{_encode_path_value(feature_id)}",
        body=feature,
        extra_headers={"Content-Type": "application/geo+json"},
    )
    return Feature.from_dict(data)


def delete_item(
    client: RoteiroClient,
    collection_id: str,
    feature_id: str,
) -> None:
    """Delete a feature from a collection.

    Args:
        client: An initialised RoteiroClient instance.
        collection_id: The collection identifier.
        feature_id: The feature identifier.
    """
    client._delete(
        f"/collections/{_encode_path_value(collection_id)}/items/{_encode_path_value(feature_id)}"
    )

File: typescript/src/collections.ts
/**
 * OGC API Features collections and feature operations.
 *
 * Standalone functions that operate on a {@link RoteiroClient} instance.
 * @module collections
 */

import type { RoteiroClient } from './client';
import type { Collection, Feature, FeatureCollection, QueryParams } from './types';

/**
 * List all OGC API collections.
 *
 * @param client - An initialised RoteiroClient instance.
 * @returns A list of Collection descriptors.
 */
export async function listCollections(client: RoteiroClient): Promise<Collection[]> {
  const resp = await client.request<{ collections: Collection[] }>('/collections');
  return resp.collections ?? [];
}

/**
 * Get a single collection by ID.
 *
 * @param client - An initialised RoteiroClient instance.
 * @param collectionId - The collection identifier.
 * @returns A Collection object.
 */
export async function getCollection(
  client: RoteiroClient,
  collectionId: string,
): Promise<Collection> {
  return client.request<Collection>(
    `/collections/${encodeURIComponent(collectionId)}`,
  );
}

/**
 * Query features from a collection with optional filters.
 *
 * @param client - An initialised RoteiroClient instance.
 * @param collectionId - The collection identifier.
 * @param params - Optional query parameters (bbox, bboxCRS, crs, limit, filter, datetime, offset).
 * @returns A FeatureCollection with matching features.
 */
export async function getItems(
  client: RoteiroClient,
  collectionId: string,
  params?: QueryParams,
): Promise<FeatureCollection> {
  const sp = new URLSearchParams();
  if (params?.bbox) sp.set('bbox', params.bbox);
  if (params?.bboxCRS) sp.set('bbox-crs', params.bboxCRS);
  if (params?.crs) sp.set('crs', params.crs);
  if (params?.limit) sp.set('limit', String(params.limit));
  if (params?.filter) sp.set('filter', params.filter);
  if (params?.datetime) sp.set('datetime', params.datetime);
  if (params?.offset !== undefined) sp.set('offset', String(params.offset));
  const q = sp.toString();
  return client.request<FeatureCollection>(
    `/collections/${encodeURIComponent(collectionId)}/items${q ? `?${q}` : ''}`,
  );
}

/**
 * Get a single feature from a collection.
 *
 * @param client - An initialised RoteiroClient instance.
 * @param collectionId - The collection identifier.
 * @param featureId - The feature identifier.
 * @returns A Feature object.
 */
export async function getItem(
  client: RoteiroClient,
  collectionId: string,
  featureId: string,
): Promise<Feature> {
  return client.request<Feature>(
    `/collections/${encodeURIComponent(collectionId)}/items/${encodeURIComponent(featureId)}`,
  );
}

/**
 * Create a new feature in a collection.
 *
 * @param client - An initialised RoteiroClient instance.
 * @param collectionId - The collection identifier.
 * @param feature - A GeoJSON Feature object.
 * @returns The created Feature.
 */
export async function createItem(
  client: RoteiroClient,
  collectionId: string,
  feature: Feature,
): Promise<Feature> {
  return client.request<Feature>(
    `/collections/${encodeURIComponent(collectionId)}/items`,
    {
      method: 'POST',
      headers: { 'Content-Type': 'application/geo+json' },
      body: JSON.stringify(feature),
    },
  );
}

/**
 * Update an existing feature in a collection.
 *
 * @param client - An initialised RoteiroClient instance.
 * @param collectionId - The collection identifier.
 * @param featureId - The feature identifier.
 * @param feature - A GeoJSON Feature object with updated data.
 * @returns The updated Feature.
 */
export async function updateItem(
  client: RoteiroClient,
  collectionId: string,
  featureId: string,
  feature: Feature,
): Promise<Feature> {
  return client.request<Feature>(
    `/collections/${encodeURIComponent(collectionId)}/items/${encodeURIComponent(featureId)}`,
    {
      method: 'PUT',
      headers: { 'Content-Type': 'application/geo+json' },
      body: JSON.stringify(feature),
    },
  );
}

/**
 * Delete a feature from a collection.
 *
 * @param client - An initialised RoteiroClient instance.
 * @param collectionId - The collection identifier.
 * @param featureId - The feature identifier.
 */
export async function deleteItem(
  client: RoteiroClient,
  collectionId: string,
  featureId: string,
): Promise<void> {
  await client.del(
    `/collections/${encodeURIComponent(collectionId)}/items/${encodeURIComponent(featureId)}`,
  );
}
